### PR TITLE
Adds HttpEndpointSupplier to allow generic discovery of HTTP endpoints

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Before, users could only use spring-cloud-sleuth to integrate HTTP senders with dynamic endpoint suppliers such as consul, eureka or zookeeper. Since sleuth is discontinued, this adds a lightweight generic functionality via a new type `HttpEndpointSupplier` which can do the same sort of translation as sleuth did (e.g. substituting a placeholder URI for a real one).

This will allow those upgrading to Spring Boot 3 to continue using service discovery such as Eureka. As an additional benefit, it also allows the non-spring ecosystem the same.

There is no current plan to introduce Eureka or any other built-in endpoint discovery mechanics to this repo. However, tests will prove they can be integrated, and the types will be validated in config scenarios including Spring Boot and raw Armeria prior to merge. If users request specific discovery providers, they can raise issues and demand for them.

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
TODO: actually integrate

File: core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * passed by configuration to a real endpoint.
+ *
+ * <p>Senders should consider the special value {@link #FIXED_FACTORY} and the type {@link Fixed} to
+ * avoid dynamic lookups when constants will be returned.
+ *
+ * <p>Senders are not called during production requests, rather in time or size bounded loop, in a
+ * separate async reporting thread. Implementations that resolve endpoints via remote calls, such as
+ * from Eureka, should cache internally to avoid blocking the reporter thread on each loop.
+ *
+ * <p>Some senders, such as Armeria, may have more efficient and precise endpoint group logic. In
```
Comment (codefromthecrypt) [inline core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:16]: For this part, I might put an example in the test tree. While this will pull armeria into our deps that are analyzed with snyk etc, usually there are no CVEs that aren't resolved quickly. Also the dependency tree isn't as huge as spring boot, so it isn't as bad for laptops in hotels.

File: core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  /**
+   * Factory passed to HTTP {@link BytesMessageSender sender} builders to control resolution of the
+   * static endpoint from configuration.
+   *
+   * <p>Unless this is {@linkplain #FIXED_FACTORY}, {@linkplain #create(String)} will be deferred to
+   * the first call to {@linkplain BytesMessageSender#send(List)}.
+   *
+   * @since 3.3
+   */
+  interface Factory {
```
Comment (codefromthecrypt) [inline core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:55]: this type will be added to spring-beans so that anyone using spring can employ it via config. that's currently the only config layer we have in this repo.

**Comment by codefromthecrypt** [comment]
ok done except 3rd-party integration tests which I'll report back in brave-example.

File: okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java [hunk 3]
```diff
@@ -113,20 +116,30 @@ public static final class Builder {
...
+     * No default. The POST HttpUrl for zipkin's <a href="https://zipkin.io/zipkin-api/#/">v2 api</a>,
      * usually "http://zipkinhost:9411/api/v2/spans"
      */
     // customizable so that users can re-map /api/v2/spans ex for browser-originated traces
     public Builder endpoint(String endpoint) {
       if (endpoint == null) throw new NullPointerException("endpoint == null");
-      HttpUrl parsed = HttpUrl.parse(endpoint);
-      if (parsed == null) throw new IllegalArgumentException("invalid POST url: " + endpoint);
-      return endpoint(parsed);
+      this.endpoint = endpoint;
```
Comment (reta) [inline okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java:136]: I think the `endpoint` is not needed anymore, is it? (we could use `endpointSupplierFactory` solely):

```java
this.endpointSupplierFactory = (HttpEndpointSupplier.Constant) endpoint -> endpoint;
```

Comment (codefromthecrypt) [reply okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java:140]: I should note where we are assigning this, that we need to retain the old endpoint or else we break the former (toBuilder()) functions. will make a comment!

Comment (codefromthecrypt) [reply okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java:140]: for example, if someone changes like .. toBuilder().endpointSupplierFactory(foo).build() we don't want to erase the (config) endpoint used to create the supplier

Comment (codefromthecrypt) [reply okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java:140]: added comments and edge case tests!

Comment (reta) [reply okhttp3/src/main/java/zipkin2/reporter/okhttp3/OkHttpSender.java:140]: Got it, thanks @codefromthecrypt !

**Review by reta — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
thanks for the review @reta! since jitpack doesn't work anymore, I'm going to merge and address anything else post-commit. This significantly reduces work for integration because for example, to try this in micrometer requires their snapshots, too!

File: core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+ * @since 3.3
+ */
+public interface HttpEndpointSupplier extends Closeable {
+  /**
+   * HTTP {@link BytesMessageSender sender} builders check for this symbol, and return the input as
+   * a {@linkplain Constant} result rather than perform dynamic lookups.
+   *
+   * @since 3.3
+   */
+  Factory CONSTANT_FACTORY = new Factory() {
```
Comment (anuraaga) [inline core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:57]: Just want to confirm this is fine to be public (private interface statics are post Java 8). Otherwise, a static method constructing a package private class can be less surface

File: core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+  /**
+   * HTTP {@link BytesMessageSender senders} check for this type, and will cache its first value.
+   *
+   * @since 3.3
+   */
+  final class Constant implements HttpEndpointSupplier {
+    private final String endpoint;
+
+    public Constant(String endpoint) {
```
Comment (anuraaga) [inline core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:103]: Related to above, this looks like the "real" public API, a static method may encapsulate better

File: core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java (hunk 4 above)
Comment (codefromthecrypt) [reply core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:57]: yep I added this so there can be a default bean in spring without thinking too much (they replicate our senders)

Comment (codefromthecrypt) [reply core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:57]: anyway no problem. just was following how like in brave we have constants like this, but it isn't important to continue that.

File: core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java (hunk 5 above)
Comment (codefromthecrypt) [reply core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:103]: ok I'll move this out to a different file and make the function. no problem.

Comment (codefromthecrypt) [reply core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:103]: actually I can just hide the ctor, we still need the type public

Comment (codefromthecrypt) [reply core/src/main/java/zipkin2/reporter/HttpEndpointSupplier.java:103]: actually if I move things I have to move everything to a new type, and now the interface has to refer to things in another file to talk about behavior. this is unlike the rest of the codebase, but I'll raise a PR and you can decide.

**Comment by codefromthecrypt** [comment]
Thanks for the review @anuraaga #251  for follow-up
